pg-index-health-spring-boot-demo/src/main/java/io/github/mfvanek/pg/index/health/demo/service/DbMigrationGeneratorService.java

```diff
@@ -0,0 +1,74 @@
+ *
+ * Licensed under the Apache License 2.0
+ */
+
+package io.github.mfvanek.pg.index.health.demo.service;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.github.mfvanek.pg.checks.cluster.ForeignKeysNotCoveredWithIndexCheckOnCluster;
+import io.github.mfvanek.pg.common.maintenance.DatabaseCheckOnCluster;
+import io.github.mfvanek.pg.connection.ConnectionCredentials;
+import io.github.mfvanek.pg.connection.HighAvailabilityPgConnection;
+import io.github.mfvanek.pg.connection.HighAvailabilityPgConnectionFactory;
+import io.github.mfvanek.pg.generator.DbMigrationGenerator;
+import io.github.mfvanek.pg.index.health.demo.dto.ForeignKeyMigrationRequest;
+import io.github.mfvanek.pg.index.health.demo.dto.ForeignKeyMigrationResponse;
+import io.github.mfvanek.pg.model.PgContext;
+import io.github.mfvanek.pg.model.constraint.ForeignKey;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+import javax.sql.DataSource;
+
+@SuppressFBWarnings("EI_EXPOSE_REP2")
+@Slf4j
+@RequiredArgsConstructor
+@Service
+@Transactional
+public class DbMigrationGeneratorService {
+
+    private final HighAvailabilityPgConnectionFactory highAvailabilityPgConnectionFactory;
+    private final DataSource dataSource;
+    private final DbMigrationGenerator<ForeignKey> dbMigrationGenerator;
+
+    public ForeignKeyMigrationResponse addIndexesWithFkChecks(final ForeignKeyMigrationRequest fkMigrationRequest) {
+        final List<ForeignKey> keysBefore = getFksFromDb(fkMigrationRequest.credentials());
+        final List<String> migrations = generatedMigrations(keysBefore);
+        final List<ForeignKey> keysAfter = getFksFromDb(fkMigrationRequest.credentials());
+        if (!keysAfter.isEmpty()) {
+            throw new IllegalStateException("There should be no foreign keys not covered by the index");
+        }
+        return new ForeignKeyMigrationResponse(keysBefore, keysAfter, migrations);
+    }
+
+    public List<ForeignKey> getFksFromDb(final ConnectionCredentials credentials) {
+        final HighAvailabilityPgConnection haPgConnection = highAvailabilityPgConnectionFactory.of(credentials);
+        final DatabaseCheckOnCluster<ForeignKey> foreignKeysNotCoveredWithIndex = new ForeignKeysNotCoveredWithIndexCheckOnCluster(haPgConnection);
+        return foreignKeysNotCoveredWithIndex.check(PgContext.of("demo"));
+    }
+
+    @SuppressFBWarnings("SIL_SQL_IN_LOOP")
+    private List<String> generatedMigrations(final List<ForeignKey> foreignKeys) {
+        final List<String> generatedMigrations = dbMigrationGenerator.generate(foreignKeys);
+        log.info("Generated migrations: {}", generatedMigrations);
+        try (Connection connection = dataSource.getConnection()) {

```

Comment:
Use JdbcTemplate here

Reply:
Could you, please, give me a hint what I am doing wrong? I tried to use JdbcTemplate here, but I faced a problem. All these "create" queries executed in one transaction, when executed in a loop. So I got an exception with a message "CREATE INDEX CONCURRENTLY cannot run inside a transaction block". I tried to put these statements in separate transactions, but failed.
Variants I implemented:
1. I used PreparedStatementCallback in order to close statement after executing. But I am not sure about the method inside lambda I chose.
This variant all statements are executed, but after all a rollback occured
![image](https://github.com/user-attachments/assets/f55d22b2-8c8d-4e9b-be81-1492eb2323d1)
2. I tried to enable autocommit mode using
```
  jdbcTemplate.getDataSource().getConnection().setAutoCommit(true);
```
but it didn't help.
3. I tried to add to statements "begin;" and "commit;" commands(with \n symbols at the end), but that didn't help. I got an error message about bad SQL grammar
4. I tried to use TransactionTemplate (with autowired PlatformTransactionManager ), but I also received exception with message  "CREATE INDEX CONCURRENTLY cannot run inside a transaction block"

![image](https://github.com/user-attachments/assets/9d255811-5bb5-4792-becf-ffe7e2221870)